Add lifted/scoped versions of functions to `named` and tutorials

srush opened the pull request:
Preliminary PR for adding "lifted" versions of pyro primitives to support functions with `latent` as the first argument. 

> conditional : takes a function with latent as the first argument, and data as a strings that are namespaced by the latent that is eventually passed in. This prevents the function from observing other global variables. 

> SVI, Search, ... : These functions now take a model and guide with latent as the first arg. For now they always use a uuid as the namespace for calling both model and guide. This is stored on the object

>  Marginal: Like with conditional the `sites` argument is now namespaced by the object passed in. The returned values are now an Object not a string dictionary that has been un-namespaced (maybe this should not be an Object)

srush commented:
You can view the new tutorials here 

https://github.com/harvardnlp/pyro/tree/sugar/tutorial/source/contrib

fritzo commented on pyro/contrib/named.py:
```diff
@@ -207,3 +207,69 @@ def param(placeholder, *args, **kwargs):
+    return partial_condition
+
+
+def lift(fn, _namespace="latent", _latent=None):
```
Is this an internal helper or an external function that users might want to use? If internal, please rename to `_lift`; if external, please add a docstring.

fritzo commented on pyro/contrib/named.py:
```diff
@@ -207,3 +207,69 @@ def param(placeholder, *args, **kwargs):
     return placeholder.param(*args, **kwargs)
+
+
+@functools.wraps(pyro.condition)
```
I'm inclined to separate these new lifted helpers out into a separate module `pyro.contrib.lifted` or `pyro.contrib.scoped`, since `pyro.contrib.named` can be used independently of these new helpers.

fritzo commented:
@srush are you able to add labels to the PR (or is that restricted to owners?) We usually use the workflow of adding either the [WIP] label or the [awaiting review] label and one or two other labels.

srush commented:
Yup will do. 

- I can't add labels, unless you add me with write access. 

- Yes contributor guide is great. I will fix that part up and document.

- Mostly just wanted to check that this was an okay track, i.e. continuing to wrap primitives to make them have named versions.

srush commented:
The namespace idea was that params and variables in `model` and `guide` need to be in the same namespace, but that any other functions without access to `latent` should not see them.  So calling SVI makes a unique prefix name, and send it only to the model and guide. If you use named you can only make variables in this namespace.

Need to check that this generalizes.

srush commented:
Okay, so:

- Added a bunch of tests in test_named.py and test_scoped.py.

- Ported over `irange`, `iarange`, and `module` to named. 

- implemented `data_fn` and `sites_fn` for specifying samples through `named` objects (see tests)

- added an example of a naive bayes HMM that shows off the scoped approach. (Unfortunately don't think VI is very good on this model... How does SVI handle discrete, REINFORCE?)

srush commented:
Oh also added pretty printing of named. I found that was a really useful way to debug. You run your function and see what comes out in your latent.

eb8680 commented:
> Unfortunately don't think VI is very good on this model... How does SVI handle discrete, REINFORCE?

The most naive version does REINFORCE, but there are also options for [Rao-Blackwellization and state-dependent baselines](http://pyro.ai/examples/svi_part_iii.html) or [enumeration](http://pyro.ai/examples/ss-vae.html#Interlude:-Summing-Out-Discrete-Latents).  In this case, enumeration would be pretty slow (see #227 for a fix) and there may be issues with broadcasting for multiple latent choices.  

You could also implement a reparametrized `Concrete` distribution or change the model to have reparametrizable states and emissions.

fritzo commented on pyro/contrib/named.py:
```diff
@@ -49,37 +77,71 @@ def __getattribute__(self, key):
+        if not self._is_placeholder:
+            raise RuntimeError("Cannot .iarange_ an initialized named.Object")
+
+        # Yields both a subsampled data and an indexed latent object.
```
Interesting design choice!

fritzo replied:
This should match the behavior of `pyro.iarange`:
```py
@functools.wraps(pyro.iarange)
def iarange_(self, *args, **kwargs):
    return pyro.iarange(self._name, *args, **kwargs)  # Returns a context manager.
```
Feel free to implement a differently-named helper that does what you want, e.g.
```py
def iaenumerate_(self, *args, **kwargs):
    with pyro.iarange(self.iarange._name, *args, **kwargs) as ind:
        yield ind, self.data
```

fritzo commented on pyro/contrib/named.py:
```diff
@@ -135,6 +220,12 @@ def __setitem__(self, pos, value):
+    def irange_(self, data, *args, **kwargs):
+        # Yields both a subsampled data and an indexed latent object.
+        for d in pyro.irange(self._name + "range", data, *args, **kwargs):
+            yield d, self.add()
```
Ha ha, this is cool!

fritzo commented on pyro/contrib/named.py:
```diff
@@ -101,13 +163,36 @@ class List(list):
         self._name = name
 
+    def _expand(self):
+        ls = []
```
Could be simplified by defining a top level helper
```py
def is_named(thing):
    return isinstance(thing, (Object, List, Dict))
```
and then this function becomes
```py
return [l._expand() if is_named(l) else l for l in self]
```

fritzo commented on pyro/contrib/scoped.py:
```diff
@@ -0,0 +1,93 @@
+        else:
+            site_collector = named.Object(latent._name)
+            data_fn(site_collector)
+            lifted_data = site_collector.visit(lambda name, val, acc: acc.setdefault(name, val),
```
And if you replace `.visit()` with the `.items()` suggested above, this simplifies to
```py
lifted_data = dict(site_collector.items())
```

fritzo commented on pyro/contrib/scoped.py:
```diff
@@ -0,0 +1,93 @@
+    if sites_fn is not None:
+        site_collector = named.Object(namespace)
+        sites_fn(site_collector)
+        sites = list(site_collector.visit(lambda name, val, acc: acc.add(name),
```
simplifies to
```py
sites = [name for name, val in site_collector.items()]
```

srush commented:
Thanks @eb8680 that's really helpful. I'll have one of my students look into `Concrete` and possibly some of the recent methods. It's cool that there is already support for baselines. 

@fritzo going through the merge and changes now.

srush commented on pyro/contrib/named.py:
```diff
@@ -28,18 +30,44 @@ class Object(object):
 
-    def __str__(self):
-        return super(Object, self).__getattribute__("_name")
+    def __repr__(self):
```
My reasoning was that users shouldn't ever *have* to see the string name. Ideally you could interact through scoped and never have to know that these were strings behind the scenes.

fritzo replied:
That sounds like a reasonable principle for the `pyro.contrib.scoped` module, but I'd like to explore uses of `pyro.contrib.named` outside of `pyro.contrib.scoped`. The `str()` method is also useful for explaining how addresses are formed (in cases where they are user facing).

srush commented on examples/contrib/hmm.py:
```diff
@@ -0,0 +1,114 @@
+def main(args):
+    optim = Adam({"lr": 0.001})
+    inference = SVI(model, guide, optim, loss="ELBO")
+    data = [[0], [1], [2]] * 20
```
will clarify.

srush commented on pyro/contrib/named.py:
```diff
@@ -49,37 +77,71 @@ def __getattribute__(self, key):
+        for param_name, param_value in nn_module.named_parameters():
+            # register the parameter in the module with pyro
+            # this only does something substantive if the parameter hasn't been seen before
+            c = self
```
Do you know if there is a safe python way to just expand string paths? Maybe the best way is `eval("c.{}".format(path)).sample_()`. In theory we could make that support lists as well.

fritzo replied:
You could adapt [pyro.util.deep_getattr](https://github.com/uber/pyro/blob/c04fc931631ec9e8694def207b5ca0e432d5e501/pyro/util.py#L389). But let's avoid `eval()` in library code :wink:

srush commented on pyro/contrib/named.py:
```diff
@@ -135,6 +220,12 @@ def __setitem__(self, pos, value):
             raise RuntimeError("Cannot overwrite {}".format(name))
         super(List, self).__setitem__(pos, value)
 
+    @functools.wraps(pyro.irange)
```
Will do. 

The problem is that there are two names here. One is a tag for the "conditional independence" tag, and the other for the rvs being uses. I'm using self._name for the List being yielded and `self._name + "range"` for the pyro tag passed to irange.

fritzo replied:
Wait, this should match the behavior of `pyro.irange()`. There is no `data` argument. I think it should be
```py
@functools.wraps(pyro.irange)
def irange_(self, *args, **kwargs):
    return pyro.irange(self._name, *args, **kwargs)  # Returns an iterator.
```
which is used like
```py
latent.xs.data = List()
for i in latent.xs.ind.irange_(10):
    x = latent.xs.data.add()
    # ...do stuff with i, x...
```
That said, it's fine to add a differently-named helper that does what you want. E.g.
```
def ienumerate_(self, *args, **kargs):
    self.data = List()
    for i in self.irange.irange_(*args, **kwargs):
        yield i, self.data.add()
```

srush commented on tests/contrib/test_named.py:
```diff
@@ -0,0 +1,96 @@
+from __future__ import absolute_import, division, print_function
```
My fault. Will merge. Your tests are way better anyway...

srush commented:
Okay, think I took care of most of these. Codes is getting simpler. Current high-level issues (not sure if these are blockers):

- Need to go through more pyro functions to understand what they return. I don't want any scoped functions to return a dict, should return Object.

- Didn't want to change SVI, so little hacked around for now.  The main idea is I would like that to be able to see the environments being created by the inference algorithm. 

- My HMM example could use some work. Need to dive into baselines etc.

fritzo commented:
@srush I'm hoping to get #540 merged this week, then you can rebase this PR onto the Pyro dev branch and I'll do one more review pass before merging. Thanks for your patience!

fritzo commented:
Ok @srush we just merged #540 . Could you please rebase onto dev for one final review?

srush commented:
Sure, will do shortly.

nehalecky commented:
Hey all—happy New Year! Just caught up with the discussion in #502 and this. Great to see this type of community here, We're definitely aligned with the changes that are being proposed. Was wondering what the status was of this PR?

srush commented:
Oh, sorry I got busy with other things, and stop using Pyro. You are welcome to take it over if you are interested.

fritzo commented:
Closing. @nehalecky feel free to revive this branch and open a new PR.
